controllers/replica_set_controller.go

```diff
@@ -331,60 +289,103 @@ func (r ReplicaSetReconciler) Reconcile(ctx context.Context, request reconcile.R
 	return res, err
 }
 
-// resetStatefulSetUpdateStrategy ensures the stateful set is configured back to using RollingUpdateStatefulSetStrategyType
-// and does not keep using OnDelete
-func (r *ReplicaSetReconciler) resetStatefulSetUpdateStrategy(mdb mdbv1.MongoDBCommunity) error {
-	if !isChangingVersion(mdb) {
-		return nil
+// deployStatefulSet deploys the backing StatefulSet of the MongoDBCommunity resource.
+// The returned boolean indicates that the StatefulSet is ready.
+func (r *ReplicaSetReconciler) deployStatefulSet(mdb mdbv1.MongoDBCommunity) (bool, error) {
+	r.log.Info("Creating/Updating StatefulSet")
+	if err := r.createOrUpdateStatefulSet(mdb); err != nil {
+		return false, fmt.Errorf("error creating/updating StatefulSet: %s", err)
 	}
-	// if we changed the version, we need to reset the UpdatePolicy back to OnUpdate
-	_, err := statefulset.GetAndUpdate(r.client, mdb.NamespacedName(), func(sts *appsv1.StatefulSet) {
-		sts.Spec.UpdateStrategy.Type = appsv1.RollingUpdateStatefulSetStrategyType
-	})
-	return err
+
+	currentSts, err := r.client.GetStatefulSet(mdb.NamespacedName())
+	if err != nil {
+		return false, fmt.Errorf("error getting StatefulSet: %s", err)
+	}
+
+	r.log.Debugf("Ensuring StatefulSet is ready, with type: %s", getUpdateStrategyType(mdb))
+
+	isReady := statefulset.IsReady(currentSts, mdb.StatefulSetReplicasThisReconciliation())
+	return isReady || currentSts.Spec.UpdateStrategy.Type == appsv1.OnDeleteStatefulSetStrategyType, nil
```

Comment:
we check for the OnDelete strategy type here as we want to be considered ready in the case of a version change even if the sts itself is not ready.